core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+    if (newInput.connection && newInput.connection !== parentInput.connection) {
+      newAstNode = Blockly.ASTNode.createInputNode(newInput);
+      break;
+    } else if (fieldNode && fieldNode.getLocation() && fieldNode !== location) {
+      newAstNode = Blockly.ASTNode.createFieldNode(fieldNode.getLocation());
+      break;
+    }
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the previous input or field given a field location.
+ * @param {!Blockly.Field} location The current location of the cursor.
+ * @param {!Blockly.Input} parentInput The parent input of the field.
+ * @return {?Blockly.ASTNode} The ast node holding the previous input or field.
+ * @private
+ */
+Blockly.ASTNode.prototype.findPrevForField_ = function(location, parentInput) {
+  var block = location.sourceBlock_;
+  var inputs = block.inputList;
+  var curIdx = inputs.indexOf(parentInput);
+  var prevLocation = this.findPreviousEditableField_(location, parentInput);
+  var newAstNode;
+
+  if (!prevLocation && curIdx - 1 >= 0) {
+    newAstNode = Blockly.ASTNode.createInputNode(inputs[curIdx - 1]);
+  }
+  return newAstNode;
+};
+
+/**
+ * Navigate between stacks of blocks on the workspace.
+ * @param {?Boolean} forward True to go forward. False to go backwards.
+ * @return {?Blockly.BlockSvg} The first block of the next stack.
+ * @private
+ */
+Blockly.ASTNode.prototype.navigateBetweenStacks_ = function(forward) {
+  var curLocation = this.getLocation();
+  if (!(curLocation instanceof Blockly.Block)) {
+    curLocation = curLocation.sourceBlock_;
+  }
+  if (!curLocation) {
+    return null;
+  }
+  var curRoot = curLocation.getRootBlock();
+  var topBlocks = curRoot.workspace.getTopBlocks();
+  for (var i = 0; i < topBlocks.length; i++) {
+    var topBlock = topBlocks[i];
+    if (curRoot.id == topBlock.id) {
+      var offset = forward ? 1 : -1;
+      var resultIndex = i + offset;
+      if (resultIndex == -1) {
+        resultIndex = topBlocks.length - 1;
+      } else if (resultIndex == topBlocks.length) {
+        resultIndex = 0;
+      }
+      return topBlocks[resultIndex];
+    }
+  }
+  throw Error('Couldn\'t find ' + (forward ? 'next' : 'previous') +
+      ' stack?!?!?!');
+};
+
+/**
+ * Create an ast node from the top connection on the block.
+ * @param {Blockly.Block} block The block that we want to find the top
+ *     connection on.
+ * @return {?Blockly.ASTNode} The astnode containing the top connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findTopASTConnection_ = function(block) {
+  var previousConnection = block.previousConnection;
+  var outputConnection = block.outputConnection;
+  var newAstNode;
+  if (previousConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(previousConnection);
+  } else if (outputConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(outputConnection);
+  }
+  return newAstNode;
+};
+
+
+/**
+ * Given a location in a stack of blocks find the next out connection. If the
```

Comment:
"Given a location find the closest location that is one level out on the AST." This is not finding an output connection. ;)